src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -2468,18 +2468,27 @@ private function registerHttpClientConfiguration(array $config, ContainerBuilder
             $retryOptions = $scopeConfig['retry_failed'] ?? ['enabled' => false];
             unset($scopeConfig['retry_failed']);
 
+            $httpClientTransport = new Reference('http_client.transport');
+
+            if ($responseFactoryId = $scopeConfig['mock_response_factory'] ?? null) {
+                $container->register('http_client.mock_client.'.$name, MockHttpClient::class)
+                    ->setDecoratedService($httpClientTransport, null, -10)  // lower priority than TraceableHttpClient (5)
+                    ->setArguments(
+                        true === $responseFactoryId ? [] : [new Reference($responseFactoryId)]);
+            }
+
             if (null === $scope) {
                 $baseUri = $scopeConfig['base_uri'];
                 unset($scopeConfig['base_uri']);
 
                 $container->register($name, ScopingHttpClient::class)
                     ->setFactory([ScopingHttpClient::class, 'forBaseUri'])
-                    ->setArguments([new Reference('http_client.transport'), $baseUri, $scopeConfig])
+                    ->setArguments([$httpClientTransport, $baseUri, $scopeConfig])
```

Comment:
not sure this change is needed. Sure you save a few objects and memory but only during build time and this is like 3 objects.